compiler/ETA/CodeGen/Bind.hs

```diff
@@ -67,23 +67,29 @@ closureCodeBody _ id lfInfo args arity body fvs binderIsFV recIds = do
     defineMethod $ mkMethodDef thisClass [Public] "getArity" [] (ret jint)
                  $  iconst jint (fromIntegral arity)
                  <> greturn jint
+    let layoutArgFieldType (_, _, ft) = ft
+        bodyType = contextType : map (layoutArgFieldType . mkLayoutArg) args
     _ <- withMethod [Public] "enter" [contextType] void $ do
       n <- peekNextLocal
       let (argLocs, code, n') = mkCallEntry n args
-          (_ , cgLocs) = unzip argLocs
-      emit code
+      emit $ gload thisFt 0 <> code
       setNextLocal n'
       bindArgs argLocs
+      mapM_ bindFV fvLocs
+      emit $ invokevirtual (mkMethodRef thisClass "body" bodyType void)
+    _ <- withMethod [Public] "body" bodyType void $ do
```

Comment:
Shouldn't this be a static method defined in `modClass`? I think this explains why you're not seeing any speedups :) It's an instance method defined in an inner class of `modClass`. The name of the static method should be identical to name of the closure. And `closureCodeBody` is used for both top-level and non top-level functions so you need to check that there are no free variables before applying the optimisation.